File: 20. 有效的括号/Solution.cpp
#include<bits/stdc++.h>

using namespace std;


class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        string lbrackets = "([{";
        string rbrackets = ")]}";
        for(int i = 0; i < s.size(); i++){
            if(lbrackets.find(s[i]) != string::npos){
                st.push(s[i]);
            }
            else if(rbrackets.find(s[i]) != string::npos){
                if(st.empty()) return false;
                char t = st.top();
                st.pop();
                if(s[i] == ')' && t == '(') continue;
                else if(s[i] == ']' && t == '[') continue;
                else if(s[i] == '}' && t == '{') continue;
                else return false;
            }
        }
        return st.empty();
    }
};

File: 20. 有效的括号/Solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for i in s:
            if i in "([{":
                stack.append(i)
            elif i in ")]}":
                if not stack:
                    return False
                t = stack.pop()
                if i == ')' and t == '(':
                    continue
                if i == ']' and t == '[':
                    continue
                if i == '}' and t == '{':
                    continue
                else:
                    return False
        return True if not stack else False